File: collect_app/src/main/java/org/odk/collect/android/widgets/ExArbitraryFileWidget.java
package org.odk.collect.android.widgets;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.util.TypedValue;
import android.view.View;

import org.javarosa.form.api.FormEntryPrompt;
import org.odk.collect.android.databinding.ExArbitraryFileWidgetAnswerBinding;
import org.odk.collect.android.formentry.questions.QuestionDetails;
import org.odk.collect.android.utilities.ApplicationConstants;
import org.odk.collect.android.utilities.QuestionMediaManager;
import org.odk.collect.android.widgets.utilities.FileRequester;
import org.odk.collect.android.widgets.utilities.WaitingForDataRegistry;

@SuppressLint("ViewConstructor")
public class ExArbitraryFileWidget extends BaseArbitraryFileWidget {
    ExArbitraryFileWidgetAnswerBinding binding;

    private final FileRequester fileRequester;

    public ExArbitraryFileWidget(Context context, QuestionDetails questionDetails,
                                 QuestionMediaManager questionMediaManager, WaitingForDataRegistry waitingForDataRegistry,
                                 FileRequester fileRequester) {
        super(context, questionDetails, questionMediaManager, waitingForDataRegistry);
        this.fileRequester = fileRequester;
    }

    @Override
    protected View onCreateAnswerView(Context context, FormEntryPrompt prompt, int answerFontSize) {
        binding = ExArbitraryFileWidgetAnswerBinding.inflate(((Activity) context).getLayoutInflater());
        setupAnswerFile(prompt.getAnswerText());

        binding.exArbitraryFileButton.setTextSize(TypedValue.COMPLEX_UNIT_DIP, answerFontSize);
        binding.exArbitraryFileAnswerText.setTextSize(TypedValue.COMPLEX_UNIT_DIP, answerFontSize);

        if (questionDetails.isReadOnly()) {
            binding.exArbitraryFileButton.setVisibility(GONE);
        } else {
            binding.exArbitraryFileButton.setOnClickListener(v -> onButtonClick());
            binding.exArbitraryFileAnswerText.setOnClickListener(v -> mediaUtils.openFile(getContext(), answerFile, null));
        }

        if (answerFile != null) {
            binding.exArbitraryFileAnswerText.setText(answerFile.getName());
            binding.exArbitraryFileAnswerText.setVisibility(VISIBLE);
        }

        return binding.getRoot();
    }

    @Override
    public void clearAnswer() {
        binding.exArbitraryFileAnswerText.setVisibility(GONE);
        deleteFile();
        widgetValueChanged();
    }

    @Override
    public void setOnLongClickListener(OnLongClickListener listener) {
        binding.exArbitraryFileButton.setOnLongClickListener(listener);
        binding.exArbitraryFileAnswerText.setOnLongClickListener(listener);
    }

    @Override
    protected void showAnswerText() {
        binding.exArbitraryFileAnswerText.setText(answerFile.getName());
        binding.exArbitraryFileAnswerText.setVisibility(VISIBLE);
    }

    @Override
    protected void hideAnswerText() {
        binding.exArbitraryFileAnswerText.setVisibility(GONE);
    }

    private void onButtonClick() {
        waitingForDataRegistry.waitForData(getFormEntryPrompt().getIndex());
        fileRequester.launch((Activity) getContext(), ApplicationConstants.RequestCodes.EX_ARBITRARY_FILE_CHOOSER, getFormEntryPrompt());
    }
}

File: collect_app/src/main/java/org/odk/collect/android/widgets/ExArbitraryFileWidget.kt
package org.odk.collect.android.widgets

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.util.TypedValue
import android.view.View
import org.javarosa.form.api.FormEntryPrompt
import org.odk.collect.android.databinding.ExArbitraryFileWidgetBinding
import org.odk.collect.android.formentry.questions.QuestionDetails
import org.odk.collect.android.utilities.ApplicationConstants
import org.odk.collect.android.utilities.QuestionMediaManager
import org.odk.collect.android.widgets.utilities.FileRequester
import org.odk.collect.android.widgets.utilities.WaitingForDataRegistry

@SuppressLint("ViewConstructor")
class ExArbitraryFileWidget(
    context: Context,
    questionDetails: QuestionDetails,
    questionMediaManager: QuestionMediaManager,
    waitingForDataRegistry: WaitingForDataRegistry,
    private val fileRequester: FileRequester,
    dependencies: Dependencies
) : BaseArbitraryFileWidget(
        context,
        questionDetails,
        questionMediaManager,
        waitingForDataRegistry,
        dependencies
    ) {
    lateinit var binding: ExArbitraryFileWidgetBinding

    override fun onCreateAnswerView(context: Context, prompt: FormEntryPrompt, answerFontSize: Int): View {
        binding = ExArbitraryFileWidgetBinding.inflate((context as Activity).layoutInflater)
        setupAnswerFile(prompt.answerText)

        binding.exArbitraryFileAnswerText.setTextSize(
            TypedValue.COMPLEX_UNIT_DIP,
            answerFontSize.toFloat()
        )

        if (questionDetails.isReadOnly) {
            binding.exArbitraryFileButton.visibility = GONE
        } else {
            binding.exArbitraryFileButton.setOnClickListener { onButtonClick() }
            binding.exArbitraryFileAnswerText.setOnClickListener {
                mediaUtils.openFile(
                    getContext(),
                    answerFile!!,
                    null
                )
            }
        }

        if (answerFile != null) {
            binding.exArbitraryFileAnswerText.text = answerFile!!.name
            binding.exArbitraryFileAnswerText.visibility = VISIBLE
        }

        return binding.root
    }

    override fun clearAnswer() {
        binding.exArbitraryFileAnswerText.visibility = GONE
        deleteFile()
        widgetValueChanged()
    }

    override fun setOnLongClickListener(listener: OnLongClickListener?) {
        binding.exArbitraryFileButton.setOnLongClickListener(listener)
        binding.exArbitraryFileAnswerText.setOnLongClickListener(listener)
    }

    override fun showAnswerText() {
        binding.exArbitraryFileAnswerText.text = answerFile!!.name
        binding.exArbitraryFileAnswerText.visibility = VISIBLE
    }

    override fun hideAnswerText() {
        binding.exArbitraryFileAnswerText.visibility = GONE
    }

    private fun onButtonClick() {
        waitingForDataRegistry.waitForData(formEntryPrompt.index)
        fileRequester.launch(
            (context as Activity), ApplicationConstants.RequestCodes.EX_ARBITRARY_FILE_CHOOSER,
            formEntryPrompt
        )
    }
}
